models/state.go

```diff
@@ -0,0 +1,104 @@
+package models
+
+//!!! this will all need a re-work for topic-api
+
+import "github.com/ONSdigital/dp-topic-api/apierrors"
+
+// State - iota enum of possible topic states
+type State int
+
+// Possible values for a State of a topic. It can only be one of the following:
+const (
+	StateCreated State = iota
+	StateUploaded
+	StateImporting
+	StateImported
+	StatePublished
+	StateCompleted
+	StateDeleted
+	StateFailedImport
+	StateFailedPublish
+	StateTrue  // to be removed at some point
+	StateFalse // to be removed at some point
+)
+
+var stateValues = []string{"created", "uploaded", "importing", "imported", "published", "completed", "deleted", "failed_import", "failed_publish", "true", "false"}
+
+// String returns the string representation of a state
+func (s State) String() string {
+	return stateValues[s]
+}
+
+// ParseState returns a state from its string representation
+func ParseState(stateStr string) (State, error) {
+	for s, validState := range stateValues {
+		if stateStr == validState {
+			return State(s), nil
+		}
+	}
+	return -1, apierrors.ErrTopicInvalidState
+}
+
+// TransitionAllowed returns true only if the transition from the current state and the provided targetState is allowed
+func (s State) TransitionAllowed(target State) bool {
```

Comment:
[This is an opinion not a strict request for change - happy to discuss and get feedback from the wider team]

This code is functionally fine - it does what its supposed to do however, its quite verbose and duplicates the same logic multiple times. If we ever add more states of change the rules for  valid state transitions it might be quite hard to maintain.

I wonder if it's possible distill this logic to be common enough to reduce the amount of repetition required. He's a noddy/over simplified example of what I am thinking:
https://play.golang.org/p/cWNNyrNp1_1

Happy to discuss.

Reply:
I will implement/test code in link provided and report back

Reply:
Done.
Code now uses State Transition Table.
See check in details.